fccodegen/_printer.py

```diff
@@ -0,0 +1,400 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+#
+import sympy
+import sympy.printing
+
+from sympy.printing.precedence import precedence
+
+#from sympy.core import S, Rational, Pow, Basic, Mul
+#from .printer import Printer
+#from sympy.printing.precedence import PRECEDENCE
+#import mpmath.libmp as mlib
+#from mpmath.libmp import prec_to_dps
+#from sympy.utilities import default_sort_key
+
+# List of functions handled by python's `math` module
+_math_functions = {
+    'acos',
+    'acosh',
+    'asin',
+    'asinh',
+    'atan',
+    'atan2',
+    'atanh',
+    'ceil',     # called ceiling in sympy
+    'cos',
+    'cosh',
+    'exp',
+    'expm1',
+    'factorial',
+    'floor',
+    'log',
+    'log10',
+    'log1p',
+    'log2',
+    'sin',
+    'sinh',
+    'tan',
+    'tanh',
+}
+
+
+class WebLabPrinter(sympy.printing.printer.Printer):
+    """
+    Converts Sympy expressions to strings for use in Web Lab Cython model code.
+
+    To use, create a :class:`WebLabPrinter` instance, and call its method
+    :meth:`doprint()` with a Sympy expression argument.
+
+    Arguments:
+
+    ``symbol_function``
+        A function that converts symbols to strings (variable names).
+    ``derivative_function``
+        A function that converts derivatives to strings.
+
+    """
+
+    def __init__(self, symbol_function=None, derivative_function=None):
+        super(WebLabPrinter, self).__init__(None)
+
+        # Prefix for functions
+        self._prefix = 'math.'
+
+        # Symbol and derivative handling (default)
+        if symbol_function is None:
+            self._symbol_function = lambda x: str(x)
+        else:
+            self._symbol_function = symbol_function
+
+        if derivative_function is None:
+            self._derivative_function = lambda x: str(x)
+        else:
+            self._derivative_function = derivative_function
+
+    def _bracket(self, expr, parent_precedence):
+        """
+        Converts ``expr`` to string, and adds parentheses around the resut, if
+        and only if ``precedence(expr) < parent_precedence``.
+        """
+        if precedence(expr) < parent_precedence:
+            return '(' + self._print(expr) + ')'
+        return self._print(expr)
+
+    def _bracket_args(self, args, parent_precedence):
+        """
+        Applies :meth:`_bracket()` to a list of expressions, and joins them
+        with a comma.
+        """
+        return ', '.join([self._bracket(x, parent_precedence) for x in args])
+
+    def emptyPrinter(self, expr):
+        """
+        Called by :class:`Printer` as a last resort for unknown expressions.
+        """
+        raise RuntimeError(
+            'Unsupported expression type (' + str(type(expr)) + '): '
+            + str(expr))
+
+    def _print_Add(self, expr):
+        """ Handles addition & subtraction, with n terms. """
+        # This method is mostly copied from sympy.printing.Str
```

Comment:
Re 'mostly' - what's different?

Reply:
There's some order preserving thing that we don't have, and the variable names are slightly different, plus some if vs ternary stuff at the bottom. I've changed it to say it's based on sympy's one

Reply:
Also I've added in comments to explain what the different bits of code do!

I always like it if the comments indicate what the code below is supposed to do (that way you can check if it does + it's more readable)